R/transform.R

```diff
@@ -0,0 +1,161 @@
+    }
+
+    trans_out <- Transforms(insertions = Insertions(data=trans$insertions),
+                            categories = NULL,
+                            elements = NULL)
+    return(trans_out)
+}
+
+#' @rdname Transforms
+#' @export
+setMethod("transforms", "CrunchVariable", getTransforms)
+#' @rdname Transforms
+#' @export
+setMethod("transforms", "VariableTuple", getTransforms)
+
+#' @rdname Transforms
+#' @export
+setMethod("transforms<-", c("CrunchVariable", "ANY"), function (x, value) {
+    frmt <- wrapEntity("view" = list("transform" = value))
+    crPATCH(self(x), body=toJSON(frmt))
+    dropCache(cubeURL(x))
+    invisible(x)
+})
+
+#' @rdname Transforms
+#' @export
+setMethod("transforms<-", c("CrunchVariable", "NULL"), function (x, value) {
+    frmt <- wrapEntity("view" = list("transform" = emptyObject()))
+    crPATCH(self(x), body=toJSON(frmt))
+    dropCache(cubeURL(x))
+    invisible(x)
+})
+
+setValidity("Transforms", function (object) {
+    one_of_names <- c("insertions", "categories", "elements") %in% names(object)
+    if (!any(one_of_names)) {
+        val <- paste("Transforms must have at least one of", serialPaste(dQuote(c("insertions", "categories", "elements")), "or"))
+    } else {
+        val <- TRUE
+    }
+
+    if (!is.null(object[["insertions"]]) && !is.insertions(object[["insertions"]])) {
+        val <- "Transforms insertions element must be of class Insertions"
+    }
+    if (!is.null(object[["categories"]])) {
+        val <- "Transforms categories element must be NULL"
+    }
+    if (!is.null(object[["elements"]])) {
+        val <- "Transforms elements element must be NULL"
+    }
+
+    return(val)
+})
+
+#' Show the variable transformations on a Categorical variable
+#'
+#' @param x a Categorical variable
+#'
+#' @return summary of the variable, with transforms applied
+#'
+#' @aliases showTransforms
+#'
+#' @examples
+#' \dontrun{
+#' showTransforms(ds$variable)
+#' }
+#'
+#' @export
+setMethod("showTransforms", "CategoricalVariable", function (x) {
+    tab <- table(x)
+    tab <- as.array(calcTransforms(tab, transforms(x), categories(x)))
+
+    remove_these <- names(categories(x)[is.na(categories(x))])
+    tab <- tab[!(names(tab) %in% remove_these)]
+    styles <- transformStyles(transforms(x), categories(x)[!is.na(categories(x))])
+
+    out <- prettyPrint2d(tab, row_styles = styles)
+    cat(unlist(out), sep="\n")
+    return(invisible(tab))
+})
+
+
+#' Test if an abstract category object is of a conceptual type
+#'
+#' Convenience functions to test if an abstract category (AbsCat) object is a
+#' specific type. These types are defined by the properties of the abstract
+#' category object.
+#'
+#' `is.abscat.subtotal` is `x` a subtotal insertion?
+#' `is.abscat.heading` is `x` a heading insertion?
+#' `is.abscat.category` is `x` a category?
+#'
+#' @param x an AbsCat object
+#'
+#' @return logical if the AbsCat object is
+#'
+#' @name AbsCat-type-tests
+#' @keywords internal
```

Comment:
Why do these exist? Shouldn't `is.Subtotal` etc. be what you use? Seems like if you're going to do introspection of the objects to infer their type, you're not getting much from using S4 classes.

Reply:
I've refactored how these works, and now everything respect s4 classes / inheritance.